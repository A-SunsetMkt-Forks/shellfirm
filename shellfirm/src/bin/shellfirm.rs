mod cmd;
use std::process::exit;

use anyhow::anyhow;
use console::{style, Style};
use shellfirm::Config;

const DEFAULT_ERR_EXIT_CODE: i32 = 1;

fn main() {
    let app = cmd::default::command()
        .subcommand(cmd::command::command())
        .subcommand(cmd::config::command());

    let matches = app.clone().get_matches();

    let env = env_logger::Env::default().filter_or(
        "LOG",
        matches.value_of("log").unwrap_or(log::Level::Info.as_str()),
    );
    env_logger::init_from_env(env);

    // load configuration
    let config = match Config::new(None) {
        Ok(config) => config,
        Err(err) => {
            eprintln!("Loading config error: {}", err);
            exit(1)
        }
    };

    let settings = match config.get_settings_from_file() {
        Ok(c) => c,
        Err(e) => {
            eprintln!(
                "Could not load setting from file. Try resolving by running `{}`\nError: {}",
                style("shellfirm config reset").bold().italic().underlined(),
                e
            );
            exit(1)
        }
    };

    let checks = match settings.get_active_checks() {
        Ok(c) => c,
        Err(e) => {
            eprintln!("Could not load checks. err: Error: {}", e);
            exit(1)
        }
    };

    let res = match matches.subcommand() {
        None => Err(anyhow!("command not found")),
        Some(tup) => match tup {
            ("pre-command", subcommand_matches) => {
                cmd::command::run(subcommand_matches, &settings, &checks)
            }
            ("config", subcommand_matches) => cmd::config::run(subcommand_matches, &config),
            _ => unreachable!(),
        },
    };

    let exit_with = match res {
        Ok(cmd) => {
            if let Some(message) = cmd.message {
                let style = if exitcode::is_success(cmd.code) {
                    Style::new().green()
                } else {
                    Style::new().red()
                };
                eprintln!("{}", style.apply_to(message));
            }
            cmd.code
        }
        Err(e) => {
            log::debug!("{:?}", e);
            DEFAULT_ERR_EXIT_CODE
        }
    };
    exit(exit_with)
}
