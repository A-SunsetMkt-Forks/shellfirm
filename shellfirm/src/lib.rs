pub mod checks;
mod config;
mod data;
pub mod dialog;
mod prompt;
pub use config::{Challenge, Config, Settings};
pub use data::CmdExit;
