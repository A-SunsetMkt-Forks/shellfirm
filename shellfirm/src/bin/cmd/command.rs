use anyhow::Result;
use clap::{Arg, ArgMatches, Command};
use lazy_static::lazy_static;
use regex::Regex;
use shellfirm::{checks, Context};

lazy_static! {
    static ref REGEX_STRING_COMMAND_REPLACE: Regex = Regex::new(r#"('|")([\s\S]*?)('|")"#).unwrap();
}

// lazy_static! {
//     static ref REGEX_STRING_COMMAND_REPLACE: HashMap<u32, &'static str> = {
//         let mut m = HashMap::new();
//         m.insert(0, "foo");
//         m.insert(1, "bar");
//         m.insert(2, "baz");
//         m
//     };
// }

pub fn command() -> Command<'static> {
    Command::new("pre-command")
        .about("Check if given command marked as sensitive command that need your extra approval.")
        .arg(
            Arg::new("command")
                .short('c')
                .long("command")
                .help("get the user command that should run.")
                .required(true)
                .takes_value(true),
        )
        .arg(
            Arg::new("test")
                .short('t')
                .long("test")
                .help("Check if the command is risky and exit")
                .takes_value(false),
        )
}

pub fn run(arg_matches: &ArgMatches, context: &Context) -> Result<shellfirm::CmdExit> {
    let command = arg_matches.value_of("command").unwrap_or(""); // todo:: wrap me

    let command = REGEX_STRING_COMMAND_REPLACE
        .replace_all(command, "")
        .to_string();

    let splitted_command: Vec<&str> = command
        .split(|c| c == '&' || c == '|' || c == "&&".chars().next().unwrap())
        .collect();

    log::debug!("splitted_command {:?}", splitted_command);
    let matches: Vec<checks::Check> = splitted_command
        .iter()
        .flat_map(|c| checks::run_check_on_command(&context.checks, c))
        .collect();

    log::debug!("matches found {}. {:?}", matches.len(), matches);
    if !matches.is_empty() {
        checks::challenge(&context.challenge, &matches, arg_matches.is_present("test"))?;
    }

    Ok(shellfirm::CmdExit {
        code: exitcode::OK,
        message: None,
    })
}
