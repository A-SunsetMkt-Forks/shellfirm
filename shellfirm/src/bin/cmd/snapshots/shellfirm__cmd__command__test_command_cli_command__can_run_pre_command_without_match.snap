---
source: shellfirm/src/bin/cmd/command.rs
expression: "run_pre_command(\"command\", &settings, &settings.get_active_checks().unwrap(),\n    true)"
---
Ok(
    CmdExit {
        code: 0,
        message: Some(
            "---\n[]\n",
        ),
    },
)
